Fix player drop handling.

pchote opened the pull request:
Fixes #10614 and possibly also the desyncs that sank #10787 and the ai-takeover PR.
Closes #10615.

pchote committed:
Remove lag-drop frame workaround.

pchote commented:
This affects more than just player lag situations (see also #5027), so I suggest this for the changelog:
* Fixed an issue that sometimes stalled multiplayer servers when a player leaves.

The easiest way to test this PR is to run multiple clients and then pause one of them by sending a `SIGSTOP` signal (`^z` in the terminal window, or `kill -SIGSTOP <pid>` from another terminal).  The client can be resumed by typing `fg` in the terminal or `kill -SIGCONT <pid>` from another.

RoosterDragon reviewed (approved):
I can confirm the bug reproduces with the above method and that this PR fixes it.

RoosterDragon commented on OpenRA.Game/Server/Connection.cs:
```diff
@@ -104,8 +104,10 @@ public void ReadData(Server server)
 						case ReceiveState.Data:
 							{
+								if (Frame > MostRecentFrame)
+									MostRecentFrame = Frame;
```
I often like to keep the variable being assigned to on the left in the conditional in expressions like this, i.e.
```c#
if (MostRecentFrame < Frame)
    MostRecentFrame = Frame;
```

That's probably just a matter of taste though.

pchote replied:
I agree and will change it.

pchote replied:
Fixed.

pchote committed:
Only update MostRecentFrame if the frame is the most recent.

This fixes it being reset to 0 after an immediate order
and not being set before dispatching (which may use it).

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/deac41cf91c70a36a9d96a448079ab54257b0843...822b82fde176bb752b53f4ec10a60bce7a6af9d9)
